libcxx/docs/ReleaseNotes/20.rst

```diff
@@ -85,6 +85,9 @@ Deprecations and Removals
 - The function ``__libcpp_verbose_abort()`` is now ``noexcept``, to match ``std::terminate()``. (The combination of
   ``noexcept`` and ``[[noreturn]]`` has special significance for function effects analysis.)
 
+- The ``_LIBCPP_DISABLE_AVAILABILITY`` macro that was used to force-disable availability markup has now been removed.
+  Whether availability markup is used by the library is now solely controlled at configuration-time.
```

Comment:
This breaks the setup in our distribution pretty essentially. We always ship an up-to-date libc++, but keep compiling for pretty old deployment targets (currently still on 10.13). The only way to make many current-gen C++ libraries work under those constraints is to disable the availability annotations, which has been how we've been supporting both modern packages AND old hardware ([docs](https://conda-forge.org/docs/maintainer/knowledge_base.html#newer-c-features-with-old-sdk)). This happens at build time for said library, but at runtime from the POV of the already-compiled libc++.

We currently have >250 [packages](https://github.com/search?q=org%3Aconda-forge%20_LIBCPP_DISABLE_AVAILABILITY&type=code) making use of this. What would be an alternative in a world after this PR? Just define `_LIBCPP_HAS_NO_VENDOR_AVAILABILITY_ANNOTATIONS`?

I realize that we're not a common use-case in and of ourselves, but we do have >2B of monthly [downloads](https://conda-forge.org/) (across all platforms, not just osx), so a relatively big multiplier. We believe that keeping compatibility for as long as feasible is an important quality, and since we've got the infrastructural wherewithal to solve the technical issues, it would be a pity to shut this door.

CC @isuruf @xhochy @beckermr @jakirkham

Reply:
If you ship your own libc++ you shouldn't enable availability annotations. I don't know exactly what kind of hacky solution you're using, but you probably want to inject your own `__config_site`. That's obviously not officially sanctioned, but should work.

Reply:
> I don't know exactly what kind of hacky solution you're using, but you probably want to inject your own `__config_site`.

If that makes it possible to tell clang to ignore the system libcxx and use the one we're shipping, that'd be great! That's really the key issue we're working around.

> If you ship your own libc++ you shouldn't enable availability annotations.

Yeah, I can understand that. Our case is simple though - our availability annotations are equivalent to what the current(==newest) version of libcxx supports, so it's equivalent to `_LIBCPP_DISABLE_AVAILABILITY` for that version AFAIU. (and there's some details w.r.t. libc++abi that I'm leaving out, see discussion in #110920).

Reply:
@h-vetinari I think what you want here is simply to define `LIBCXX_ENABLE_VENDOR_AVAILABILITY_ANNOTATIONS=OFF` at CMake configuration time. In fact, `LIBCXX_ENABLE_VENDOR_AVAILABILITY_ANNOTATIONS=OFF` is the default, so I think there's something I'm not understanding about your use case. Why do you manually enable `LIBCXX_ENABLE_VENDOR_AVAILABILITY_ANNOTATIONS=ON` and then attempt to disable it via some other mechanism?

Reply:
@ldionne I think they actually use the Apple SDK headers but ship their own libc++.so to support some programs on older platforms.

Reply:
Ugh. That's very very unsupported though. You can't use headers and a dylib from mismatched configurations, that's literally asking for problems. What if the Apple SDK headers were configured to use a different ABI configuration, for example?

@h-vetinari is that really what you're doing? I'd be willing to chat with you to understand your use case and see how it can be supported, however it seems like your current setup might not be something we can support as-is. I'm sure there's a proper way to support your use case though.

Reply:
`-nostdinc++` works just fine with the SDK, right? They should probably just use that and then `-isystem /path/to/libcxx` just like usually.

Reply:
Yes, `-nostdinc++` works when a SDK is in use. There's nothing different when a SDK is in use, really: it's just that the sysroot where Clang starts looking for stuff is located inside the SDK, but all the behavior is otherwise the same.

Reply:
So to answer the questions w.r.t. our setup (to the best of my understanding, much of the setup was done before my time, I'm trying to document & improve things as I go).

Because our users are necessarily installing a full environment (where we control dependencies and distribution), we can rely on the fact that we can always ship an up-to-date `libc++.dylib`. We also use up-to-date the libc++ headers when compiling things, but we still need to use point to _some_ Apple SDK (for which older versions are available, e.g. from https://github.com/phracker/MacOSX-SDKs or https://github.com/alexey-lysiuk/macos-sdk), and we're generally on a 10.13 baseline for now.

We need the SDK for platform-related things and the C stdlib, so this is kind of unavoidable. However, the libc++ headers will trigger availability warnings based on the detected macOS version and therefore report false positives (from our POV, because our C++ stdlib is always up-to-date and available).

These things happen at runtime (from the POV of libc++), because different packages may require a newer SDK for some reason, and we can provide that based on the configuration of a given build (often while still leaving `MACOSX_DEPLOYMENT_TARGET` unchanged).

We're also relying on ABI stability of libc++, in the sense that artefacts compiled against libc++ vN get a runtime dependency `libcxx >=N`, so we that if we later have `libc++.{N+2}.dylib` in the environment, things still work in the vast majority of situations. I don't want to scare people here in that this is what the ironclad expectation is, but rather that this is something that has worked for us for years, and rebuilding literally 10's of thousands of artefacts for every new libc++ version is simply not realistic.

So if `_LIBCPP_DISABLE_AVAILABILITY` gets removed, AFAIU we'd have to either:
* Patch things so that we recreated the effect of `_LIBCPP_DISABLE_AVAILABILITY`
* Provide our own vendor availability annotations that match that effect

Neither sounds impossible, but we try to minimize the amount of [patches](https://github.com/conda-forge/libcxx-feedstock/tree/main/recipe/patches) we carry, obviously...

Actually there is one exception where we do modify the default availability annotations, which has to do with another wrinkle. While we could build and ship `libc++abi` just fine (and we did in the past), there are situations where certain libraries or frameworks (including CPython) will bypass the fact that everything points to our custom `$PREFIX` and simply go to the system `libc++abi` straight away. Because it's basically impossible to chase down and patch all cases where this happens, we've stopped shipping `libc++abi` and now always rely on the system-provided one, which is why we unconditionally disable `_LIBCPP_AVAILABILITY_HAS_INIT_PRIMARY_EXCEPTION` for example, as the required `___cxa_init_primary_exception` isn't available before macOS 15 AFAIU (and our deployment target baseline is much lower). This is also why I'm interested in `-DLIBCXX_CXX_ABI=system-libcxxabi`, c.f. https://github.com/llvm/llvm-project/pull/110920.

Finally, where we're aware of it, we do patch relevant ABI differences for older macOS system libs, compare e.g. [this patch](https://github.com/conda-forge/libcxx-feedstock/blob/main/recipe/patches/0001-Fix-ABI-compatibility-with-system.patch) to realign a struct, resp. the related [discussion](https://discourse.llvm.org/t/shipping-custom-libc-on-macos/58606) (I think the original email [thread](https://lists.llvm.org/pipermail/libcxx-dev/2021-July/001172.html) got split a bit by the discourse migration). This isn't great, but given that the system ABI isn't 100% stable, this is +/- the best we can do under the given constraints.

Hope that explains the situation a bit more. Happy to answer more questions of course, and thanks for giving us a chance to make the case!

Reply:
@h-vetinari All of this doesn't explain why you can't use your own libc++ headers via `-nostdinc++`. Is it just that you've never though of that or is there an issue with using that strategy?

Reply:
Thanks a lot for the comprehensive response here. It was a fair amount for me to chew on, which is why I ended up postponing having to deal with it a few times - apologies 😅 

I'll try to use "just the C bits" of the SDK as you describe. It sounds like a more solid approach already so I'm hoping that'll work. 🤞 

Also 💯  on the different symbol namespaces, that would have been excellent. Finally, the case you commented on in discourse really is unusual (also for us), in it's triggered by an uncommon case of flags (using `-Wl,-rpath`, but not `-L` to point to `$PREFIX/lib`, where our `libc++` lives.)